installinator/src/block_size_writer.rs

```diff
@@ -0,0 +1,227 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use pin_project_lite::pin_project;
+use std::io;
+use std::pin::Pin;
+use std::task::ready;
+use std::task::Context;
+use std::task::Poll;
+use tokio::io::AsyncWrite;
+
+pin_project! {
+    /// `BlockSizeWriter` is analogous to a tokio's `BufWriter`, except it
+    /// guarantees that writes made to the underlying writer are always
+    /// _exactly_ the requested block size, with two exceptions: explicitly
+    /// calling (1) `flush()` or (2) `shutdown()` will write any
+    /// buffered-but-not-yet-written data to the underlying buffer regardless of
+    /// its length.
+    ///
+    /// When `BlockSizeWriter` is dropped, any buffered data it's holding will
+    /// be discarded. It is critical to manually call `BlockSizeWriter:flush()`
+    /// or `BlockSizeWriter::shutdown()` prior to dropping to avoid data loss.
```

Comment:
I don't know how to feel about this, seems like a big footgun. I would prefer either writing the data in the `Drop` or just panicking if there's any data to be written.

Reply:
I don't think we can write the data without async drop, right?

I agree that this is a pretty gnarly footgun, but it's the exact same one tokio's `BufWriter` has. `std`'s BufWriter notes that it tries to write the data in drop but you have no feedback if it fails, so even if we could write the data we'd still want a "but you should really use flush when you're done" doc comment just like `std`'s BufWriter does.

Every time I've tried to add a panic in drop to catch client misbehavior it's come back to bite me. I'd be perfectly fine putting a `debug_assert` that the buffer is empty in `Drop`, but I'm not sure that would actually help us since the block writer only runs on real hardware (and is built in release mode basically always AFAIK).

Reply:
Aha, forgot about async drop. Yeah, matching tokio's `BufWriter` is ok.